### ISSUE TITLE
Problem with ewa_* scalers

### ISSUE THREAD

**Issue by Argon-** [mpv-player/mpv#4607]
### mpv version and platform

Latest version as of now (18c74f7dfe9cbefe9de818caa00c3f77ce600f71) on OSX 10.11 using a AMD GPU. 
GPU + Apple drivers should support OpenGL 4.1.  
This does not happen on a Broadwell iGPU (but otherwise identical system/configuration).

### Reproduction steps

Play video with scale/cscale set to an ewa_* scaler.

### Expected behavior

Normal playback.

### Actual behavior

Wrong colors with `cscale=ewa_*` or strange artifacts with `scale=ewa_*`.

### Log file

Bisecting identifies this one as first bad commit: https://github.com/mpv-player/mpv/commit/ad0d6caac76a0cff9e98912314e749c1fde32d98
I double checked and the commit right before this one does not exhibit the problem.

Log showing cscale issue using `--no-config --cscale=ewa_lanczos --no-resume-playback -v`: https://0x0.st/G4L.txt
Log showing scale issue using `--no-config --scale=ewa_hanning --no-resume-playback -v`: https://0x0.st/G43.txt

Screenshot of unscaled video: https://0x0.st/G4t.png
Screenshot of fullscreen video exhibiting `scale` issue: https://0x0.st/G4v.png
Screenshot of fullscreen video exhibiting `cscale` issue: https://0x0.st/G4S.png

All logs were made with a ad0d6caac76a0cff9e98912314e749c1fde32d98 build.

**Comment by haasn** [comment]
Seems like indexing is backwards on OS X for some reason. Does this diff fix it?

```diff
diff --git a/video/out/opengl/video_shaders.c b/video/out/opengl/video_shaders.c
index 83b9f83143..77d8747994 100644
--- a/video/out/opengl/video_shaders.c
+++ b/video/out/opengl/video_shaders.c
@@ -181,7 +181,7 @@ void pass_sample_polar(struct gl_shader_cache *sc, struct scaler *scaler,
                 // Gather the four surrounding texels simultaneously
                 for (int n = 0; n < components; n++) {
                     GLSLF("c%d = textureGatherOffset(tex, base, ivec2(%d, %d), %d);\n",
-                          n, x, y, n);
+                          n, x, y, components - 1 - n);
                 }
 
                 // Mix in all of the points with their weights
```

**Comment by Argon-** [comment]
No, looks just as wrong

**Comment by haasn** [comment]
Screenshot? What about `3 - n`?

**Comment by haasn** [comment]
Hmm no even if I convert to ycbcr, swap the chroma channels, and convert back, it looks way different. And in the `scale` image, it looks like it exhibits weird, grid-like artifacts

**Comment by Argon-** [comment]
All with `--no-config --scale=ewa_hanning --no-resume-playback`:

`components - 1 - n`: https://0x0.st/GJG.png
`3 - n`: https://0x0.st/GJd.png

**Comment by haasn** [comment]
Well I'm pretty much out of ideas. There must be some severe bug going on with this platform/driver combination:

1. It seems like the sample order is wrong
2. Even ignoring the sample order, it's as if the different planes get somehow mixed up

What's more annoying is that I can't seem to figure out in what way the plane mix-up is happening. Anything obvious doesn't seem to solve it, and I can't even manually figure out how to “undo” the error (e.g. by swapping the chroma planes).

It would probably help if you used some more regular test pattern like this image:

![test](https://0x0.st/Gyu.jpg)

**Comment by haasn** [comment]
Here, this one's better:

![colortest](https://0x0.st/Gy1.jpg)

**Comment by Argon-** [comment]
`--no-config --scale=ewa_hanning --window-scale=2 --keep-open --no-border`:
Current HEAD: https://0x0.st/GyL.png
`components - 1 - n`: https://0x0.st/Gyp.png
`3 - n`: https://0x0.st/Gyf.png

**Comment by haasn** [comment]
> Current HEAD: https://0x0.st/GyL.png

Have you tried refilling your monitor's color ink?

**Comment by haasn** [comment]
The results of some testing over IRC:

```
00:45 <hanna> so it breaks as soon as you go from textureGather to textureGatherOffset?
00:45 <hanna> even with offset = 0?
```

**Comment by haasn** [comment]
Narrowing it down some more: Seems like `textureGatherOffset` with `int component` != `0` breaks, even though `textureGather` works fine.

**Comment by haasn** [comment]
Anyway, this is clearly adriver bug, since that trivial shader breaks even though the usage is perfectly within the scope of the `textureGatherOffset` specification, and furthermore it works fine for `textureGather`.

Also works fine for other platforms/drivers.

**Comment by Akemi** [comment]
i just tested it on 10.12 on an NVIDIA GPU with the official NVIDIA Drivers, this works as expected. will test with the Apple Drivers tomorrow after a restart.

**Comment by haasn** [comment]
It would be useful if somebody could test AMD+Windows/Linux

**Comment by Argon-** [comment]
It might really just be a bug only surfacing in a combination of this driver + this fairly outdated GPU. While this combination is possible, it's certainly not common at all.

**Comment by AirPort** [comment]
> It might really just be a bug only surfacing in a combination of this driver + this fairly outdated GPU. 
For what it's worth I checked and it also shows up on Sierra with the 2015 MBP's discrete GPU (AMD).

**Comment by Argon-** [comment]
Oh, now that's interesting because I'm using an older OS (10.11) and a way older GPU (6870).
This means this bug could cover quite a lot Macs with dedicated GPUs, as they are always AMD (Apple switched in 2014/2015 to AMD for all dedicated GPUs).

**Comment by JCount** [comment]
While I know this may not be helpful, I can also confirm the behavior on the same platform as @AirPort. I can try testing on the second public beta of High Sierra, if anyone thinks that might be useful.

**Comment by Argon-** [comment]
Yes please. This is clearly a driver bug, maybe they fixed it in 10.13.

**Comment by JCount** [comment]
The issue is still present in 10.13 Beta, unfortunately.

**Comment by haasn** [comment]
So, solution ideas?

1. Blacklist this GL version from the AMD driver in mpv code (very ugly but will work for all AMD/OS X users, difficult to tell when it will start working in the future and we should remove the blacklist)
2. Add an option to disable GLSL 400 (so users affected by this bug can just blacklist it themselves)
3. Replace EWA code by computer shader kernel (sort of WIP on my end, should be even faster than the textureGather code and will hopefully work on AMD/OSX systems)
4. Tell apple to fix it and wait (good luck)
5. Tell users to use LD_PRELOAD hacks to hide the buggy GLSL version. (Requires no change on our part)

**Comment by Argon-** [comment]
4. is unlikely
5. just for the record, it's DYLD_INSERT_LIBRARIES on OSX but not sure how your suggestion would work
3. sounds like the best solution to me considering you already work on that anyway (at least it sounds like you)

Among 1 and 2 I personally perfer 2, but that's probably not the most user-friendly solution.

**Comment by Akemi** [comment]
> 3. Replace EWA code by computer shader kernel (sort of WIP on my end, should be even faster than the textureGather code and will hopefully work on AMD/OSX systems)

correct me if i am wrong, but compute shaders are a feature of opengl 4.3 and upwards? macOS is sitting on opengl 4.1 since years and Apple is not planning to change anything in that regard. so that won't work for macOS.

**Comment by haasn** [comment]
> correct me if i am wrong, but compute shaders are a feature of opengl 4.3 and upwards? macOS is sitting on opengl 4.1 since years and Apple is not planning to change anything in that regard. so that won't work for macOS.

You can check with `glxinfo | grep compute_shader`.

**Comment by Argon-** [comment]
There is no `glxinfo` on OSX unless you install X11. However, I think it only supports a ~2.1 legacy context.
Nevertheless, Akemi is correct. Currently, Apple only supports OpenGL 4.1 and apparently computer shaders are a 4.3 feature. :/

**Comment by Akemi** [comment]
that's probably the best equivalent to glxinfo.
![screen shot 2017-07-17 at 16 39 55](https://user-images.githubusercontent.com/680386/28273785-88030a34-6b0f-11e7-9f80-09561c396f39.png)

**Comment by Argon-** [comment]
Could someone with a legitimate AMD-using Mac maybe file this as bug with Apple? Maybe, just maybe, it will get fixed. It affects most/all of their supported systems and their latest iteration/choice of graphic cards, after all...

**Comment by JCount** [comment]
This is a complete guess on my part (I have only a very basic knowledge of GLSL), but could this possibly be due to the implementation-specific behavior described here?: https://github.com/KhronosGroup/OpenGL-Registry/blob/master/extensions/ARB/ARB_texture_gather.txt#L304-L312

I can confirm the the macOS implementation of `GL_ARB_texture_gather` does indeed define
`MAX_PROGRAM_TEXTURE_GATHER_COMPONENTS_ARB`.

```
MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Versions/A/Headers/gl3ext.h
63-
64:#ifndef GL_ARB_texture_gather
65-#define GL_MIN_PROGRAM_TEXTURE_GATHER_OFFSET_ARB 0x8E5E
66-#define GL_MAX_PROGRAM_TEXTURE_GATHER_OFFSET_ARB 0x8E5F
67-#define GL_MAX_PROGRAM_TEXTURE_GATHER_COMPONENTS_ARB 0x8F9F
68-#endif
--
229-
230:#ifndef GL_ARB_texture_gather
231:#define GL_ARB_texture_gather 1
232-#endif
233-
```

I would try to check the value of this in `pass_sample_polar` myself, but unfortunately I simply do not know how. If someone would be willing to give me sample code to test, I am more than willing to do so.

**Comment by haasn** [comment]
Oh, good catch. That could totally be it. One simple way to test would be to make the following change:

```glsl
diff --git a/video/out/opengl/video_shaders.c b/video/out/opengl/video_shaders.c
index 22c1789d7f..9c89cb7060 100644
--- a/video/out/opengl/video_shaders.c
+++ b/video/out/opengl/video_shaders.c
@@ -189,7 +189,7 @@ void pass_sample_polar(struct gl_shader_cache *sc, struct scaler *scaler,
             if (use_gather) {
                 // Gather the four surrounding texels simultaneously
                 for (int n = 0; n < components; n++) {
-                    GLSLF("c%d = textureGatherOffset(tex, base, ivec2(%d, %d), %d);\n",
+                    GLSLF("c%d = textureGather(tex, base + pt * vec2(%d, %d), %d);\n",
                           n, x, y, n);
                 }
 
```

It's also possible, however, that gathering is only supported if the number of components is below a threshold.

**Comment by JCount** [comment]
@haasn That seems to fix it! [Output](https://0x0.st/kPQ.log)
```
mpv --no-config --cscale=ewa_lanczos --pause -v /tmp/mpv_ewa/big_buck_bunny_720p_h264.mov
[vo/opengl] Initializing OpenGL backend 'cocoa'
[vo/opengl] GL_VERSION='4.1 ATI-1.51.8'
[vo/opengl] Detected desktop OpenGL 4.1.
[vo/opengl] GL_VENDOR='ATI Technologies Inc.'
[vo/opengl] GL_RENDERER='AMD Radeon R9 M370X OpenGL Engine'
[vo/opengl] GL_SHADING_LANGUAGE_VERSION='4.10'
```


<img width="955" alt="screen shot 2017-07-28 at 21 52 02" src="https://user-images.githubusercontent.com/18537310/28741359-7da11ee2-73e2-11e7-969a-7d790733a312.png">

**Comment by sss123next** [comment]
i have same problem on linux

```
OpenGL vendor string: X.Org
OpenGL renderer string: AMD CAPE VERDE (DRM 2.49.0 / 4.9.40, LLVM 6.0.0)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 17.3.0-devel (git-06e20c4b8c)
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:

```

will check suggested patch later.

**Comment by sss123next** [comment]
patch does not help on linux

![mpv_cscale](https://user-images.githubusercontent.com/202772/28741544-efbd0e4e-7420-11e7-8b0d-469b226722d4.png)
![mpv_scale](https://user-images.githubusercontent.com/202772/28741547-f7ac0060-7420-11e7-8835-f2d3721025b9.png)

**Comment by grumpbot** [comment]
I have the same problem as @sss123next.  It doesn't happen on the 0.26 release but it does happen on the 0.26.0_46_gdf12f76910 version given by the AUR mpv-git package.

**Comment by haasn** [comment]
@JCount does this commit work for you?

https://github.com/haasn/mpv/commit/bf0f6671d38bf910a3a3df42687964c79243af48

**Comment by haasn** [comment]
@sss123next @grumpbot that's a different issue, for a different shader

**Comment by Argon-** [comment]
Fyi: the patch above makes it work for me as well (guess that was to be expected as the offset function is no longer used?) but https://github.com/haasn/mpv/commit/bf0f6671d38bf910a3a3df42687964c79243af48 doesn't, same problem.

**Comment by haasn** [comment]
@Argon- That's odd. Can you try changing the `<=` to `<`?

**Comment by haasn** [comment]
@Argon- re-reading this issue I think we'd already established that just using textureGatherOffset breaks for you, even with offset=0.

It would be useful if you could figure out the exact min/max values your driver reports.

**Comment by Argon-** [comment]
Making mpv print the values results in:

 vo/opengl: GL_MAX_PROGRAM_TEXTURE_GATHER_OFFSET: 31.
 vo/opengl: GL_MIN_PROGRAM_TEXTURE_GATHER_OFFSET: -32.

**Comment by haasn** [comment]
Oh, I see what the problem is. Sign confusion.

**Comment by haasn** [comment]
Actually, even when checking for the limit on both sides, that still won't fix it for you, because those checks will still both succeed.

**Comment by JCount** [comment]
@haasn Unfortunately, the code you provided still displays the same behavior on my discrete AMD M370X GPU.

**Comment by haasn** [comment]
@JCount including the patched version at https://github.com/haasn/mpv/commit/82f419aaee7a17f410701716b09294bbd3883222 ?

**Comment by JCount** [comment]
Yes.

**Comment by haasn** [comment]
Is there any error in the log? If not, then I doubt we'll learn anything

So it seems like your idea of checking against the min/max offset was a shot in the dark, and the problem apparently still seems to be that `textureGatherOffset` itself is fundamentally broken on whatever platform this is.

**Comment by sss123next** [comment]
should i create separated ticket for my problem ? or i can provide any additional info here ?
tell me what i need to provide.

**Comment by sss123next** [comment]
mpv log scale: https://bpaste.net/show/721458cc9464
note what problem with scale appears only if any scaling is done, in  original aspect video output is correct.

**Comment by haasn** [comment]
@sss123next Your issue should be fixed in master. (Still unrelated to this issue)

**Comment by sss123next** [comment]
thx, i will check now.
sorry for misspost
i have asked about new ticket ;)

**Comment by sss123next** [comment]
yes. my problem fixed in master.
thx.

**Comment by AirPort** [comment]
I have a legitimate AMD equipped Mac, so if anyone wants/still care I can try to fill the bug to Apple, but I think I'll need some help to write a meaningful report.

**Comment by haasn** [comment]
The bug seems to be narrowed down to `textureGatherOffset(sampler2D, ivec2(0), 1)` breaking.

**Comment by AirPort** [comment]
Could something like this be fine?

> The mpv video player's OpenGL renderer makes use of textureGatherOffset for calculations related to EWA image scalers (relevant portion of code: https://github.com/mpv-player/mpv/blob/master/video/out/opengl/video_shaders.c#L187).
> This works as expected everywhere except on AMD hardware (affecting a lot of Mac configurations out there), where colors appears to be mangled, as shown by the following test images:
> 
> Original image: https://0x0.st/Gy1.jpg
> Same image, scaled with the aforementioned code: https://0x0.st/GyL.png
> 
> The breakage seems to happen at 
> `textureGatherOffset(sampler2D, ivec2(0), 1)`
> 
> The issue seems to affect all AMD drivers at least in macOS 10.11, 10.12 and 10.13 Public Beta; with intel and nvidia drivers (both system and web ones) the code works fine.
> 
> 
> Link to the original mpv bug report: https://github.com/mpv-player/mpv/issues/4607

**Comment by haasn** [comment]
Oh the example code should be `textureGatherOffset(sampler2D, texcoord, ivec2(0), 1)`. Otherwise I guess that's fine, as long as you link to this thread so they can get more context.

**Comment by AirPort** [comment]
And sent. Let's see if they even remotely consider it.

**Comment by Argon-** [comment]
I saw this discussion in the channel (shortened):

    <Cpuroast> that's funny, ewa_lanczos messes with the colors badly on the AMD GPU
    <Cpuroast> ewa_lanczossharp does not
    <hanna> what about setting --scale=ewa_lanczos and also --scale-radius=3.0, --scale-radius=3.2 and --scale-radius=3.5
    <hanna> I mean there's a hardware capability (maximum texture gather offset) which we currently ignore
    <hanna> but that would cause issues when you start *exceeding* the maximum offset
    <hanna> as opposed to when you don't
    <Cpuroast> 3.0 breaks
    <Cpuroast> 3.2 and 3.5 does not
    <hanna> what about 2.0, 4.0?
    <hanna> what about 2.9?
    <Cpuroast> 2.0 and 4.0 are good
    <Cpuroast> 2.9 breaks
    <hanna> so there's a narrow range in which it breaks, but everything outside that range works?!
    <hanna> Cpuroast: https://github.com/mpv-player/mpv/issues/4607 so is this not the same issue?
    <hanna> that one broke even for offset=0
    <Cpuroast> 2.1 to 3.1 breaks
    <Cpuroast> 3.2 and up works
    <hanna> https://0x0.st/nLO.txt
    <hanna> Cpuroast: that's suspiciously “regular”
    <hanna> try diffing the generated shaders for 2.09 and 2.11 or something
    <hanna> maybe there's an obvious change
    <hanna> http://opengl.gpuinfo.org/gl_stats_caps_single.php?listreportsbycap=GL_MAX_PROGRAM_TEXTURE_GATHER_OFFSET
    <hanna> Cpuroast: oh hang on, if you set --scale-cutoff to 0, does that make it 2.0 and 3.0 exactly?
    <hanna> it might be because radius_cutoff < radius
    <hanna> (although even with --scale-cutoff=0, we cut off the kernel on actual 0 parameters)
    <Cpuroast> Anything below scale radius 1.0 also fails
    <Cpuroast> 1.0 and below
    <Cpuroast> 1.1 to 2.0 works
    <Cpuroast> Haven't tried scale-cutoff yet
    <Cpuroast> 9.0 works, 10.0 fails
    <Cpuroast> So some scale radius values it likes others it doesn't
    <Cpuroast> 6.0 and 7.0 yes, 8.0, no, 9.0 yes, 10.0 no


I can confirm this, certain values actually do work. However, for me it's not exactly the same as mentioned by Cpuroast and they are also different for different scalers (e.g. ewa_hanning doesn't work with radius 3.2 while ewa_lanczos does). `scale-cutoff` does influence it but doesn't make all values work.
As I didn't saw the requested diff performed:
1. doesn't work: `--no-config --pause --window-scale=1.5 --scale=ewa_lanczos --scale-radius=3.1` https://0x0.st/n5o.txt
2. works: `--no-config --pause --window-scale=1.5 --scale=ewa_lanczos --scale-radius=3.2` https://0x0.st/n5H.txt
3. Diff: https://0x0.st/n5X.txt (I removed the `[vo/opengl] [  1] ` parts prior to diffing)

At a quick glance, only looking at the `textureGatherOffset` calls, it looks like the working shader is only using `ivec2` with `|1|` values.

**Comment by JCount** [comment]
There is also apparently an implementation-dependent `GL_MIN_PROGRRAM_TEXTURE_GATHER_OFFSET`.

> It perfoms as textureGather but with offset applied as described in textureOffset, except that the implementation-dependent minimum and maximum offset values are given by GL_MIN_PROGRRAM_TEXTURE_GATHER_OFFSET and GL_MAX_PROGRAM_TEXTURE_GATHER_OFFSET, respectively.

https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/textureGatherOffset.xhtml

**Comment by haasn** [comment]
@JCount Careful: That's for the minimum *offset* (i.e. negative values), not the minimum *distance*. It's basically just the inverse of _MAX_OFFSET. For example, on my platform, the max offset is 31 and the min offset is -32.

Incidentally, a theory for why it breaks on only “even” sizes: We only use the gather code for contiguous groups of 4, and otherwise fall back to “direct sampling”. So for some sizes (which should be odd sizes in theory), it will use some `textureGather()` calls and some `texture()` calls - whereas for others, it will only use `textureGather()` calls. It would be helpful if you could identify which case breaks, and which doesn't - i.e., does it work if you use both, or does it only work if you use one or the other?

For a hare-brained shot in the dark, it's possible that the OpenGL driver's memory dependency resolver only looks at `texture()` calls to figure out if a texture is being sampled from. That would explain why it works as soon as you insert a `texture()`, but breaks when you're only using `textureGather()`. These are AMD GPUs, right? I know AMD uses image layouts heavily, so a wrong image layout (from a missing image barrier) could very well completely fuck the image as we see.

If my guess is correct, then it's possible that we could work around this by inserting a dummy texture() call to make sure the driver gets the dependency right. If so, this patch would fix it:

```diff
diff --git a/video/out/opengl/video_shaders.c b/video/out/opengl/video_shaders.c
index a7b253249a..7af9660762 100644
--- a/video/out/opengl/video_shaders.c
+++ b/video/out/opengl/video_shaders.c
@@ -163,6 +163,7 @@ static void polar_sample(struct gl_shader_cache *sc, struct scaler *scaler,
 void pass_sample_polar(struct gl_shader_cache *sc, struct scaler *scaler,
                        int components, int glsl_version)
 {
+    GLSL(color = texture(tex, pos);)
     GLSL(color = vec4(0.0);)
     GLSLF("{\n");
     GLSL(vec2 fcoord = fract(pos * size - vec2(0.5));)
```

**Comment by Argon-** [comment]
Unfortunately, that didn't do it. I checked the output and the line was not eliminated:

    [vo/opengl] [  1] #undef tex
    [vo/opengl] [  2] #undef texmap
    [vo/opengl] [  3] #define tex texture0
    [vo/opengl] [  4] #define texmap texmap0
    [vo/opengl] [  5] vec2 pos = texcoord0;
    [vo/opengl] [  6] vec2 size = texture_size0;
    [vo/opengl] [  7] vec2 pt = pixel_size0;
    [vo/opengl] [  8] color = texture(tex, pos);
    [vo/opengl] [  9] color = vec4(0.0);
    [vo/opengl] [ 10] {
    [vo/opengl] [ 11] vec2 fcoord = fract(pos * size - vec2(0.5));
    ...

**Comment by haasn** [comment]
@Argon- That's mpv's output, not the compiled shader. Could try removing the line that says “color = vec4(0.0);” as well. (And probably also add a “0.01 * ” in front of the “texture()”)

**Comment by Argon-** [comment]
Oh, I somehow assumed it only performs optimization based on this representation. Is there a way to check the compiled output?

**Comment by wm4** [comment]
I doubt the shader compiler is dumb enough not to remove dead stores.

**Comment by Argon-** [comment]
Tried this (and also each change on its own) but no luck:

    -    GLSL(color = vec4(0.0);)
    +    GLSL(color = 0.01 * texture(tex, pos);)
    +    //GLSL(color = vec4(0.0);)

**Comment by JCount** [comment]
I can report that using textureGatherOffset (the ewa_* scalers) seems to have been fixed for me by https://github.com/mpv-player/mpv/commit/b940691.

Platform: macOS 10.12.6
GPU: AMD Radeon R9 M370X

**Comment by AirPort** [comment]
I have the same hardware and the same system but I still can reproduce the issue even with the latest master.

**Comment by JCount** [comment]
Ok, I seem to have been overly broad in my statement. It seems to have specifically fixed `ewa_lanczossharp`.

**Comment by Argon-** [comment]
This has always been the case but somehow it isn't reported here (at least I can't see it). Certain scale radii do work, e.g. `ewa_lanczossharp` works but ewa_hanning with the same radius doesn't.

**Comment by haasn** [comment]
@Argon- It is reported. In fact, by you

**Comment by JCount** [comment]
So am I just blindly running into the fact that `ewa_lanczossharp` works with this given radius and wrongly attributing it to a random commit that happens to contain some modification of the relevant code like a complete idiot? (Be honest)

**Comment by Argon-** [comment]
Does this commit really make a difference? Because when it's only modifying the compute shader version, then you are not even using it. OSX doesn't support compute shaders, unfortunately.

**Comment by AirPort** [comment]
Suprisingly enough, a year later and despite Apple's deprecation of OpenGL in macOS, a dev actually replied to my bug report

> This link no longer works: https://github.com/mpv-player/mpv/blob/master/video/out/opengl/video_shaders.c#L187
> 
> Can you point to a specific filter with the issue and the code path/shaders that it uses?
> 
> We have a number of internal tests using textureGatherOffset() and they run fine, would it be possible to provide a simpler test case to reproduce the erroneous behavior?
> 
> Also it's worth noting that on newer OS versions (macOS 10.14 beta build 18A333 in my case) the app crashes immediately. I'm going to try building it myself now as the public drops could be built on older frameworks causing the incompatibility.
> 
> Can you please verify links, and provide isolation details on which filter?

I replied with the updated link to the code, and reported the observation about some radius' values working. Is there anything I can add to the reply in order for it to be more useful?

**Comment by Argon-** [comment]
@AirPort the crash he experienced is probably what will be fixed once PR #5953 is merged. This is indeed a 10.14 problem

**Comment by varenc** [comment]
This old bug is still occurring in macOS 10.15.2.  The `--scale-radius=3.2` trick as a workaround still works.

The old screenshot links also seem to be broken, so here's some new screenshots for comparison.

---

### `mpv --no-config`
![image](https://user-images.githubusercontent.com/167585/79273127-5ade6400-7e57-11ea-9198-b88c9e3702d6.png)


---

### `mpv --no-config --scale=ewa_lanczos  --scale-radius=3.1`

![image](https://user-images.githubusercontent.com/167585/79273189-7184bb00-7e57-11ea-911e-a23425afe633.png)

---

### `mpv --no-config --scale=ewa_lanczos  --scale-radius=3.2`
![image](https://user-images.githubusercontent.com/167585/79273224-819c9a80-7e57-11ea-8142-bed2d82bf19c.png)

**Comment by Argon-** [comment]
I'm quite certain Apple will never fix this bug. They probably haven't touched their opengl code for quite some time and see no reason to ever touch it again.

**Comment by Akemi** [comment]
i can still reproduce this on my AMD GPU. it was reported to Apple and Apple has since then deprecated opengl. so we will never see a direct fix for this, other than maybe a opengl to metal compatibility layer similar to iOS. i can also confirm that this problem doesn't occur on my WIP moltenvk/vulkan branch. 

because of the aforementioned reasons i will closes this issue.
